_posts/2016-04-26-aws-and-consul.md

```diff
@@ -0,0 +1,175 @@
+
+### User Data
+If we build this now, we will get a set of five Consul nodes, we won't get a Cluster. As I alluded to before, we need to do a little post launch config to get the nodes to start talking to each other. If we make sure our Consul AMI has the AWS CLI tools installed, it's fairly easy to use that to query AWS and ask for all instances tagged 'consul', get their IP addresses, shove those values into Consul's config file and finally, start the service.
+
+```
+#!/bin/bash
+
+internalIP=$(curl http://169.254.169.254/latest/meta-data/local-ipv4)
+instanceID=$(curl http://169.254.169.254/latest/meta-data/instance-id)
+hostname="consul-${instanceID#*-}"
+
+hostnamectl set-hostname $hostname
+
+aws ec2 describe-instances --region eu-west-1 --filters 'Name=tag:Name,Values=consul' 'Name=instance-state-name,Values=running' | jq -r '.Reservations[].Instances[].PrivateIpAddress' > /tmp/instances
+
+while read line;
+do
+ if [ "$line" != "$internalIP" ]; then
+    echo "Adding address $line"
+    cat /etc/consul/000-consul.json | jq ".retry_join += [\"$line\"]" > /tmp/${line}-consul.json
+
+    if [ -s /tmp/${line}-consul.json ]; then
+        cp /tmp/${line}-consul.json /etc/consul/000-consul.json
+    fi
+ fi
+done < /tmp/instances
+
+rm -f /tmp/instances
+
+# Clear any old state from the build process
+rm -rf /var/lib/consul/*
+
+systemctl start consul
+```
+**Nb.** If using this script as a Terraform template, don't forget to escape ${blah} with a second dollar or Terraform will try to interpolate these variables itself. 
+
+## Pulling It All Together
+We now have all the elements we need. Let's describe from start to finish what happens during a rolling upgrade of the cluster.
+
+ 1. Produce a new AMI with your upgrades installed
+ 2. Change the AMI value in your launch configuration to point at the new image
+ 3. New ASG+LC is created alongside the existing cluster
+ 4. As instances are launched, they query the AWS API to discover the IP addresses of the existing cluster
+ 5. Consul server starts and data starts replicating to new nodes
+ 6. Load balancer health checks pass for the new nodes
+ 7. Destruction of old nodes initiated, each one gracefully leaving the cluster
+ 8. New leader elected
+ 9. Profit
+
+An important point to note is that if at the ELB health checks do not pass during a given timeout, the new instances are destroyed without ever taking control.
+
+## Summary
+I hope during this whistle-stop tour of AWS, Consul (and a little Terraform) you have learned some interesting concepts that you can apply to your own projects. This pattern of performing outage-less upgrades does not just apply to Consul of course. In theory, any server deployment which has clustering or even just has some mechanism of storing the data off-host, can with a little effort be deployed to an Auto Scaling Group and gain all the advantages that go along with it.
```

Comment:
I don't feel like this bit reads well: "... off-host, can with a little effort be deployed to an Auto Scaling Group and gain all the advantages that go along with it".

maybe: "... off-host, can, with a little effort, be deployed....."?